tools/user/reorg/redirects/renames-to-redirects.js

```diff
@@ -0,0 +1,168 @@
+const basePath = path.resolve(__dirname, "../../../..");
+const baseBranch = args[0] || "origin/develop";
+
+
+// first bit of this script is synchronous - there's absolutely nothing we can do until git gives us the list of renames,
+// and I don't feel like dealing with the crufty old exec() interface
+let branch = "";
+let renames = [];
+try {
+  branch = execSync("git rev-parse --abbrev-ref HEAD", {cwd: basePath}).toString().trim();
+  renames = execSync(`git diff --name-status ${baseBranch} ${branch}`, {cwd: basePath}).toString()
+    .split('\n')
+    .filter(l => l.startsWith("R") && l.endsWith(".mdx"))
+    .map(l => l.split('\t').slice(1));
+} catch (e) {
+  throw new Error("Error running git: " + e.toString());
+}
+
+
+// this does everything else: 
+// - creates a map of current path names to redirects based on the git-identified renames 
+// - reads each mdx file in turn
+// - finds an appropriate redirect if possible 
+// - adds, updates or removes a redirect for this branch 
+// - writes out the file (if it has changed)
+// A comment in the frontmatter YAML is used to associate *at most one* redirect with the current branch.
+// This script will only touch redirects tagged with a comment for the current branch name
+const run = async () => {
+
+  const fsPathToURLPath = (fsPath) => {
+    // 1. strip leading product_docs/docs and advocacy_docs
+    // 2. strip trailing index.mdx
+    // 3. strip trailing .mdx
+    // 4. strip trailing / 
+    return fsPath.replace(/^advocacy_docs|^product_docs\/docs|index\.mdx$|\.mdx$/g, '').replace(/\/$/, '');
+  };
+
+  const mapCurrentPathToRedirect = {};
+  for (const [before, after] of renames)
+  {
+    const oldUrlPath = fsPathToURLPath(before);
+    const newUrlPath = fsPathToURLPath(after);
+    const redirectPath = path.relative(newUrlPath, oldUrlPath);
+    mapCurrentPathToRedirect[after] = redirectPath;
+  }
+
+  const sourceFiles = await glob([path.resolve(basePath, "product_docs/**/*.mdx"), path.resolve(basePath, "advocacy_docs/**/*.mdx")]);
+
+  let found = 0, updated = 0;
+  for (const sourcePath of sourceFiles)
+  {
+    const redirectPath = mapCurrentPathToRedirect[path.relative(basePath, sourcePath)];
+    if (redirectPath)
+      found++;
+    const origContent = await readFile(sourcePath, "utf8");
+    const newContent = await replaceRedirect(origContent, redirectPath, {sourcePath, branch});
+    if (newContent !== origContent )
+    {
+      updated++;
+      await writeFile(sourcePath, newContent, "utf8");
+    }
+  }
+
+  console.log(`${renames.length} renames identified, ${found} redirects identified, ${updated} files updated`);
+
+
+};
+
+const replaceRedirect = async (content, redirectPath, {sourcePath, branch}) => {
+  const frontmatterRE = /^(?<open>---\s*?\n)(?<yaml>.+?\n)(?<close>---\s*?(?:\n|$))/s;
+  const tagComment = "generated for " + branch;
+
+  const frontmatterYaml = content.match(frontmatterRE)?.groups?.yaml;
```

Comment:
For future reference, I've found https://www.npmjs.com/package/front-matter simplifies the front-matter processing no end.

Reply:
The oddly complicated approach here is purely so that I can *write the frontmatter back out* with as few disturbances as possible - ideally, the only changes are those I *intend* to make, although I'm not overly worried about changing indentation if I'm making other modifications. In particular, I don't want to mess up comments, which most normal YAML parsers throw away (as they should).